File: src/main/java/leetcode/hot100/LeetCode5.java
package leetcode.hot100;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/3 14:23
 */
public class LeetCode5 {

    public static void main(String[] args) {
        System.out.println(new LeetCode5().longestPalindrome("abba"));
        System.out.println(new LeetCode5().longestPalindrome("ss"));
        System.out.println(new LeetCode5().longestPalindrome("s"));
        System.out.println(new LeetCode5().longestPalindrome("aba"));
        System.out.println(new LeetCode5().longestPalindrome("babad"));
    }

    String res = "";
    String temp;

    public String longestPalindrome(String s) {

        if (s == null || s.length() == 0) {
            return "";
        }

        if (s.length() == 1) {
            return s;
        }

        for (int i = 0; i < s.length() - 1; i++) {
            find(s, i, i);
            if (temp.length() > res.length()) {
                res = temp;
            }
            find(s, i, i + 1);
            if (temp.length() > res.length()) {
                res = temp;
            }

        }
        return res;

    }

    // "abba"
    //   12
    //  0  3
    private void find(String s, int i, int j) {
        if (s.charAt(i) != s.charAt(j)) {
            return;
        }
        while (i > 0 && j < s.length() - 1 && s.charAt(i - 1) == s.charAt(j + 1)) {
            i--;
            j++;
        }
        temp = s.substring(i, j + 1);
    }
}

File: go/src/leetcode/hot100/LeetCode5.go
package hot100

func longestPalindrome(s string) string {
	if len(s) == 0 {
		return s
	}
	var res string
	for i := range s {
		a := cal(s, i, i)
		b := cal(s, i, i+1)
		if len(a) > len(res) {
			res = a
		}
		if len(b) > len(res) {
			res = b
		}
	}
	return res
}

func cal(s string, i, j int) string {
	if j > len(s)-1 || s[i] != s[j] {
		return ""
	}
	for (i >= 1) && j < len(s)-1 {
		if s[i-1] == s[j+1] {
			i--
			j++
		} else {
			break
		}
	}
	return s[i : j+1]
}
